src/ast/nodes/shared/Node.ts

```diff
@@ -161,10 +168,12 @@ export class NodeBase extends ExpressionEntity implements ExpressionNode {
 		super();
 		// Nodes can opt-in to keep the AST if needed during the build pipeline.
 		// Avoid true when possible as large AST takes up memory.
-		this.esTreeNode = keepEsTreeNode ? esTreeNode : null;
-		this.keys = keys[esTreeNode.type] || getAndCreateKeys(esTreeNode);
+		if (keepEsTreeNode) {
+			this.esTreeNode = esTreeNode;
+		}
+		ensureKeysAreDefinedForNodeType(esTreeNode);
```

Comment:
The reason it was written the way it was written before was to avoid the function call in the 99% case where we already have the keys. But if your measurements indicate that there was no performance degradation, we can also do it this way.

Reply:
That's fair. We can still write this in a way that avoids the function call, but doesn't assign to a member variable.

Reply:
```ts
const {type} = esTreeNode;
if (!keys[type]) {
	keys[type] = createKeysForNode(esTreeNode);
}
```

Reply:
You could also write it as

```js
keys[type] ||= createKeysForNode(esTreeNode);
```

though I assume the engine will produce the same code.